bench/micro/Bench/Database/LSMTree/Internal/WriteBuffer.hs

```diff
@@ -0,0 +1,323 @@
+import           System.Directory (removeDirectoryRecursive)
+import qualified System.FS.API as FS
+import qualified System.FS.IO as FS
+import qualified System.FS.IO.Internal.Handle as FS (HandleOS (..))
+import           System.IO.Temp
+import qualified System.Random as R
+import           System.Random (StdGen, Uniform, mkStdGen, uniform, uniformR)
+
+benchmarks :: Benchmark
+benchmarks = bgroup "Bench.Database.LSMTree.Internal.WriteBuffer" [
+      benchWriteBuffer configWord64
+        { name         = "word64-2k"
+        , ninserts     = 2_000
+        }
+    , benchWriteBuffer configWord64
+        { name         = "word64-10k"
+        , ninserts     = 10_000
+        }
+    , benchWriteBuffer configWord64
+        { name         = "word64-50k"
+        , ninserts     = 50_000
+        }
+    , benchWriteBuffer configUTxO
+        { name         = "utxo-2k"
+        , ninserts     = 1_000
+        , ndeletes     = 1_000
+        }
+    , benchWriteBuffer configUTxO
+        { name         = "utxo-10k"
+        , ninserts     = 5_000
+        , ndeletes     = 5_000
+        }
+    , benchWriteBuffer configUTxO
+        { name         = "utxo-50k"
+        , ninserts     = 25_000
+        , ndeletes     = 25_000
+        }
+    , benchWriteBuffer configUTxO
+        { name         = "utxo-2k-blob"
+        , ninserts     =   500
+        , nblobinserts =   500
+        , ndeletes     = 1_000
+        }
+    , benchWriteBuffer configUTxO
+        { name         = "utxo-10k-blob"
+        , ninserts     = 2_500
+        , nblobinserts = 2_500
+        , ndeletes     = 5_000
+        }
+    , benchWriteBuffer configUTxO
+        { name         = "utxo-50k-blob"
+        , ninserts     = 12_500
+        , nblobinserts = 12_500
+        , ndeletes     = 25_000
+        }
+    ]
+
+benchWriteBuffer :: Config -> Benchmark
+benchWriteBuffer conf@Config{name} =
+    withEnv $ \ ~(_dir, hasFS, kops) ->
+      bgroup name [
+          bench "insert" $
+            Cr.nf (\kops' -> insert kops') kops
+        , Cr.env (pure $ insert kops) $ \wb ->
+            bench "flush" $
+              Cr.perRunEnvWithCleanup getPaths (const (cleanupPaths hasFS)) $ \p ->
+                closeRun hasFS =<< flush hasFS p wb
+        , bench "insert+flush" $
+            -- To make sure the WriteBuffer really gets recomputed on every run,
+            -- we'd like to do: `Cr.nfAppIO (kops' -> ...) kops`.
+            -- However, we also need per run cleanup to avoid running out of
+            -- disk space, e.g. using `perRunEnvWithCleanup`.
+            -- Unfortunately, both take an IO action and return `Benchmarkable`,
+            -- so we can't combine them.
+            -- As a workaround, we thread `kops` through the environment, too.
+            Cr.perRunEnvWithCleanup
+              ((,) kops <$> getPaths)
+              (const (cleanupPaths hasFS)) $ \(kops', p) ->
+                closeRun hasFS =<< flush hasFS p (insert kops')
```

Comment:
This is where it gets annoying, since `perRunEnv` uses `NFData` and if I don't want to use it then the recursive directory delete becomes part of the benchmarked code. Not sure if there's a way to work around it except copying internals out of `criterion`.

Reply:
I just return `()` now and force the run to WHNF myself, as suggested.